src/main/java/games/strategy/engine/data/GameParser.java

```diff
@@ -732,11 +732,25 @@ private void parseConnections(final List<Element> connections) throws GameParseE
     }
   }
 
-  private void parseResources(final Element root) {
-    getChildren("resource", root).stream()
-        .map(e -> e.getAttribute("name"))
-        .map(name -> new Resource(name, data))
-        .forEach(data.getResourceList()::addResource);
+  private void parseResources(final Element root) throws GameParseException {
+    for (final Element element : getChildren("resource", root)) {
+      final String name = element.getAttribute("name");
+      final String isDisplayedFor = element.getAttribute("isDisplayedFor");
+      final List<PlayerID> players = data.getPlayerList().getPlayers();
```

Comment:
nit: unnecessary `data` access here.  This is only needed ```final List<PlayerID> players = data.getPlayerList().getPlayers();``` if the boolean conditional is false. 

This block could be made a bit simpler. I see a problem that `players` is initialized before it is needed. Thinking about flow, we grab XML `isDisplayedData`, we grab players list from game data, then we check `isDisplayedData` and if none then we clear the player list we just grabbed : )

So to be specific the two problems I see are:
- `players` variable is initialized before it is needed, outside of the scope of if block
- `players` variable is re-used for a second purpose (SRP violation). That would be fixed by removing the `players.clear` calls and simply creating new lists as needed.

Combining the above with some other suggestions I had along the way, this block can be simplified to at least the following:
```
if(element.getAttribute("name")
              .getAttribute("isDisplayedFor")
               .equalsIgnoreCase("NONE")) {
   data.getResourceList().addResource(new Resource(name, data));
} else {
  final List<PlayerID> players = data.getPlayerList().getPlayers();
  data.getResourceList().addResource(new Resource(
     name, 
     data,  
     isDisplayedFor.split(":").stream()
        .map(s -> data.getPlayerList().getPlayerId(s))
        .collect(Collectors.toList());
}
```

Personally, I'd like to see it go a bit further even, that such logic is moved and fully contained to a new class (strategy pattern). In this way we could test the new logic much more readily, even if we can't yet test all of GameParser.

Reply:
I really think that is much harder to read and maintain than the original.

Reply:
The original, it's hard to tell where ` final List<PlayerID> players = data.getPlayerList().getPlayers();` is used. In the updated version it's clear it's in the else block. Another iteration can probably go further and do it all as one block. Regardless, there is the concern that the original and updated are hard to maintain and lacking tests.

Reply:
These are still problems if you sue streaming notation or not:
```
- players variable is initialized before it is needed, outside of the scope of if block
- players variable is re-used for a second purpose (SRP violation). That would be fixed by removing the players.clear calls and simply creating new lists as needed.
```

```
+      if (isDisplayedFor.equalsIgnoreCase(RESOURCE_IS_DISPLAY_FOR_NONE)) {
        data.getResourceList().addResource(new Resource(name, data));
+      } else if (!isDisplayedFor.isEmpty()) {
          List<PlayerId> playersWithResource = new ArrayList<>();
+        for (final String s : isDisplayedFor.split(":")) {
+          final PlayerID player = data.getPlayerList().getPlayerId(s);
+          if (player == null) {
+            throw new GameParseException("Parse resources could not find player: " + s);
+          }
+          playersWithResource.add(player);
+        }
         data.getResourceList().addResource(new Resource(name, data, players));
+      }
+    }
```
That woudl fix SRP and make it a bit less direct. The single void call at the end I don't think is a good thing, it requires tracing the 'players' object through the above if/else/for block, it's not very direct.

Reply:
I can see what you are getting at just isn't that easy to make it much simpler. I refactored it to instead not re-use players list across if conditions. I'd say its about the same in complexity/readability but it essentially comes down to if avoiding the re-use and clearing of players list is worth having an extra if and 3 calls to addResource. Its probably more explicit this way.